### PR TITLE
Support global_phase in `ReflectionUsingPrepare`

### PR DESCRIPTION
**Pull request by tanujkhattar**
Different papers in the literature use differen versions of the "Reflection" operator that have slightly different global phases. It's convenient to make the global phase part of the signature so that controlled reflection bloqs (often used) have the right behavior. 

Some examples of papers that use different global phases:
1. https://arxiv.org/abs/1805.03662 - Qubitization formula uses $R_{s} = (2|s\rangle \langle s| - I)$ whereas the circuit decomposition is for $R_{s} = (I - 2|s\rangle \langle s|)$
2. https://arxiv.org/abs/2401.10321 - $R_{s} = i (2|s\rangle \langle s| - I)$

Pulled out from https://github.com/quantumlib/Qualtran/pull/816

### CODE REVIEW

File: qualtran/bloqs/reflection_using_prepare_test.py [hunk 1]
```diff
@@ -86,11 +86,12 @@ def get_3q_uniform_dirac_notation(signs):
 
 @pytest.mark.parametrize('num_ones', [*range(5, 9)])
 @pytest.mark.parametrize('eps', [0.01])
-def test_reflection_using_prepare(num_ones, eps):
+@pytest.mark.parametrize('global_phase', [+1, -1])
```
Comment (fdmalone) [inline qualtran/bloqs/reflection_using_prepare_test.py:89]: does 1j work?

File: qualtran/bloqs/reflection_using_prepare.py [hunk 2]
```diff
@@ -30,12 +30,17 @@
...
-    Applies $R_{s} = I - 2|s><s|$ using $R_{s} = P(I - 2|0><0|)P^{\dagger}$ s.t. $P|0> = |s>$.
+    Applies $R_{s, g=1} = g (I - 2|s><s|)$ using $R_{s} = P(I - 2|0><0|)P^{\dagger}$
+    s.t. $P|0> = |s>$.
+
     Here
         $|s>$: The state along which we want to reflect.
         $P$: Unitary that prepares that state $|s>$ from the zero state $|0>$
         $R_{s}$: Reflection operator that adds a `-1` phase to all states in the subspace
             spanned by $|s>$.
+        $g$: The global phase to control the behavior of the reflection. For example:
```
Comment (fdmalone) [inline qualtran/bloqs/reflection_using_prepare.py:41]: Maybe write this down like $R = g (I - 2|s><s|$)?

**Review by fdmalone — CHANGES_REQUESTED** [review]
This is an improvement. Minor questions.

File: qualtran/bloqs/reflection_using_prepare_test.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/bloqs/reflection_using_prepare_test.py]: Yes, added tests for `1j` and `-1j`

File: qualtran/bloqs/reflection_using_prepare.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/reflection_using_prepare.py:41]: done

**Comment by tanujkhattar** [comment]
@fdmalone Thanks for the review, I've addressed your comments.

**Review by fdmalone — APPROVED** [review]

**Comment by mpharrigan** [comment]
This slowed the test way down. Please mark as slow and include a quick version

**Pull request merged**